### PR TITLE
#2 initial Server And Get + Caching System

### LINKED ISSUE #2: Enhance learning Session With Asynchronous Content Loading
To increase the scope of the learning session we want to try teaching async operations within swift and completing them using swiftui. in order to do this we want to implement a very basic nodejs server with an endpoint that will return a list of cards that the app will have to render. We might also want to implement creating a new card from the app and uploading it to the server as well. All Server work will be completed, only leave gaps on app side.

### PR DESCRIPTION
**Pull request by Peyton-McKee**
Added Basic Typescript Node Server for Retrieving data when not stored on device, waits five seconds to exemplify the effects of caching and the waiting process for the user. 

Created an AsyncContentLoader component with a loadable object protocol to implement to manage the state of an async object

made deck a loadable object along with the loaded props for the cards. 

Adjusted the coding keys for card to make id optional

made deck conform to codable protocol

probably some other changes im forgetting

Ideally well have them implement adding or removing flashcards as an activity to the server. if you agree ill add that funcitonality to this pr (i.e. a sqlite db if we want to store between sessions with the associated routes and such on the backend and then template functions for them to fill out to talk with the server.)

https://github.com/Oasis-NEU/multi-flashcards/assets/113635669/3ac8938d-b1a9-419a-934b-3c7b8052d1cb

closes #2

### CODE REVIEW

File: multi-flashcards/APIHandler.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import Foundation
+
+final class APIHandler {
+    static let baseURL = "http://localhost:3000"
+    
+    static func getDeck() async throws -> Deck {
+        let url = URL(string: Self.baseURL + "/deck")!
+        let (data, _) = try await URLSession.shared.data(from: url)
+        print(String(data: data, encoding: .utf8))
```
Comment (fraander) [inline multi-flashcards/APIHandler.swift]: Replace this with some sort of error handling/error message?

File: multi-flashcards/AsyncContentLoader/AsyncContentView.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import SwiftUI
+
+struct AsyncContentView<Source: LoadableObject, Content: View>: View {
+    @ObservedObject var source: Source
+    var content: (Source.Output) -> Content
+        
+    var body: some View {
+        switch source.state {
+        case .idle:
+            Color.clear
```
Comment (fraander) [inline multi-flashcards/AsyncContentLoader/AsyncContentView.swift]: This feels like an anti-pattern ...
Could you wrap the switch in a Group and put a task on the group?

**Review by fraander — APPROVED** [review]
Couple small questions, but overall really sweet!

File: multi-flashcards/APIHandler.swift (hunk 1 above)
Comment (Peyton-McKee) [reply multi-flashcards/APIHandler.swift]: my bad that was just for debugging purposes forgot to delete

File: multi-flashcards/AsyncContentLoader/AsyncContentView.swift (hunk 2 above)
Comment (Peyton-McKee) [reply multi-flashcards/AsyncContentLoader/AsyncContentView.swift]: asked on slack

File: backend/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    definition: "True",
+  },
+];
+
+const deck = { cards: cards };
+
+app.use(cors());
+app.use(express.json());
+
+app.get("/", (_: Request, res: Response) => {
```
Comment (fraander) [inline backend/src/index.ts:36]: You _could_ put this all in sep. files as bEsT pRaCtIcE but yeah this all looks good

File: backend/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      res.status(500).send("An unknown error occurred");
+    }
+  }
+});
+
+app.delete("/card/:id", (req: Request, res: Response) => {
+  try {
+    const { id } = req.params;
+
+    const cardId = parseInt(id, 10);
```
Comment (fraander) [inline backend/src/index.ts:69]: by default it's base-10 so the ,10 is not needed 🤷

**Review by fraander — APPROVED** [review]

**Pull request merged**